### PR TITLE
Blobstore live server workaround

### PR DESCRIPTION
**Pull request by ChenyuHeidiZhang**
In this PR I did the workaround to make blobstore work on live server. I switched to storing blobkey in the datastore instead of imageUrl and rendering image with a separate servlet

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java [hunk 1]
```diff
@@ -95,17 +90,17 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
     String mood = request.getParameter("mood");
     String content = request.getParameter("comment-content");
 
-    // Get the URL of the image uploaded by the user to Blobstore.
-    String imageUrl = getUploadedFileUrl(request, "image");
-
+    // Get the blobKey of the image uploaded by the user to Blobstore.
```
Comment (hayleejane3) [inline portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java]:
```suggestion
    // Get the BlobKey of the image uploaded by the user to Blobstore.
```

File: portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java [hunk 2]
```diff
@@ -115,41 +110,34 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
...
+   * Returns null if the user did not upload a file.
+   */
+  private String getUploadedBlobKey(HttpServletRequest request, String formInputElementName) {
     BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
     Map<String, List<BlobKey>> blobs = blobstoreService.getUploads(request);
     List<BlobKey> blobKeys = blobs.get(formInputElementName);
-
-    // User submitted form without selecting a file, so we can't get a URL. (dev server)
+    
+    // User submitted form without selecting a file, so we can't get a blobKey. (dev server)
```
Comment (hayleejane3) [inline portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java]:
```suggestion
    // User submitted form without selecting a file, so we can't get a BlobKey. (dev server)
```

File: portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java [hunk 3]
```diff
@@ -115,41 +110,34 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
...
+   * Returns the String representation of the blobKey of the file uploaded by the user to Blobstore. 
+   * Returns null if the user did not upload a file.
+   */
+  private String getUploadedBlobKey(HttpServletRequest request, String formInputElementName) {
     BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
     Map<String, List<BlobKey>> blobs = blobstoreService.getUploads(request);
     List<BlobKey> blobKeys = blobs.get(formInputElementName);
-
-    // User submitted form without selecting a file, so we can't get a URL. (dev server)
+
```
Comment (hayleejane3) [inline portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java]: Remove extra space

File: portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java [hunk 4]
```diff
@@ -115,41 +110,34 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
     // Redirect back to the HTML page.
     response.sendRedirect("/comments.html");
   }
-
-  /** Returns a URL that points to the uploaded file, or null if the user didn't upload a file. */
-  private String getUploadedFileUrl(HttpServletRequest request, String formInputElementName) {
+  
+  /**
+   * Returns the String representation of the blobKey of the file uploaded by the user to Blobstore. 
+   * Returns null if the user did not upload a file.
```
Comment (hayleejane3) [inline portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java]: You might need to fix the wrapping, but this is how we write the `null` return cases.

```suggestion
   * Returns the String representation of the blobKey of the file uploaded by the user to Blobstore or {@code null} if 
   * the user did not upload a file.
```

File: portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java [hunk 5]
```diff
@@ -115,41 +110,34 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
...
-
-    // User submitted form without selecting a file, so we can't get a URL. (dev server)
+    
+    // User submitted form without selecting a file, so we can't get a blobKey. (dev server)
     if (blobKeys == null || blobKeys.isEmpty()) {
       return null;
     }
 
-    // Our form only contains a single file input, so get the key at the first index.
+    // Our form only contains a single file input, so get the first index.
```
Comment (hayleejane3) [inline portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java]: I think you still want to say "get the key at the first index"

File: portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java [hunk 6]
```diff
@@ -115,41 +110,34 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
...
     BlobKey blobKey = blobKeys.get(0);
+
+    // User submitted form without selecting a file, so we can't get a blobKey. (live server)
```
Comment (hayleejane3) [inline portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java]:
```suggestion
    // User submitted form without selecting a file, so we can't get a BlobKey. (live server)
```

File: portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java [hunk 7]
```diff
@@ -115,41 +110,34 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
...
     BlobInfo blobInfo = new BlobInfoFactory().loadBlobInfo(blobKey);
-    // User submitted form without selecting a file, so we can't get a URL. (live server)
     if (blobInfo.getSize() == 0) {
       blobstoreService.delete(blobKey);
       return null;
     }
-
+
```
Comment (hayleejane3) [inline portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java]: Space here too.

File: portfolio/src/main/java/com/google/sps/servlets/ServeImageServlet.java [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+import com.google.appengine.api.blobstore.BlobKey;
+import com.google.appengine.api.blobstore.BlobstoreService;
+import com.google.appengine.api.blobstore.BlobstoreServiceFactory;
+import java.io.IOException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/** Servlet that serves an image given a blobKey. */
```
Comment (hayleejane3) [inline portfolio/src/main/java/com/google/sps/servlets/ServeImageServlet.java]:
```suggestion
/** Servlet that serves an image when given a BlobKey. */
```

File: portfolio/src/main/webapp/script.js [hunk 9]
```diff
@@ -202,15 +202,20 @@ function createCommentElement(comment) {
...
+      const imageLink = document.createElement('a');
+      imageLink.href = imageUrl;
+      const imageElement = document.createElement('img');
+      imageElement.src = imageUrl;
+      imageLink.appendChild(imageElement);
+      commentContainer.appendChild(imageLink);
+    })    
   }
-
+
```
Comment (hayleejane3) [inline portfolio/src/main/webapp/script.js]: Extra Space :)

**Review by hayleejane3 — CHANGES_REQUESTED** [review]

File: portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java (hunk 1 above)
Comment (ChenyuHeidiZhang) [reply portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java]: ok

File: portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java (hunk 2 above)
Comment (ChenyuHeidiZhang) [reply portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java]: ok

File: portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java (hunk 3 above)
Comment (ChenyuHeidiZhang) [reply portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java]: ok

File: portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java (hunk 4 above)
Comment (ChenyuHeidiZhang) [reply portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java]: got it

File: portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java (hunk 5 above)
Comment (ChenyuHeidiZhang) [reply portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java]: right

File: portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java (hunk 6 above)
Comment (ChenyuHeidiZhang) [reply portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java]: ok

File: portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java (hunk 7 above)
Comment (ChenyuHeidiZhang) [reply portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java]: ok

File: portfolio/src/main/java/com/google/sps/servlets/ServeImageServlet.java (hunk 8 above)
Comment (ChenyuHeidiZhang) [reply portfolio/src/main/java/com/google/sps/servlets/ServeImageServlet.java]: ok

File: portfolio/src/main/webapp/script.js (hunk 9 above)
Comment (ChenyuHeidiZhang) [reply portfolio/src/main/webapp/script.js]: got it

File: portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java (hunk 3 above)
Comment (ChenyuHeidiZhang) [reply portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java]: wait I don't see extra space?

Comment (ChenyuHeidiZhang) [reply portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java]: Ah I see what it is - just invisible

**Review by hayleejane3 — APPROVED** [review]
Thanks :)

**Pull request merged**